diag_manager/fms_diag_axis_object.F90

```diff
@@ -1174,6 +1196,68 @@ subroutine write_diurnal_metadata(this, fileobj)
   end subroutine write_diurnal_metadata
 
+  !> @brief Creates a new z subaxis to use
+  subroutine create_new_z_subaxis(zbounds, var_axis_ids, diag_axis, naxis, file_axis_id, nfile_axis)
+    real(kind=r4_kind),                       intent(in)    :: zbounds(2)      !< Bounds of the Z axis
+    integer,                                  intent(inout) :: var_axis_ids(:) !< The variable's axis_ids
+    class(fmsDiagAxisContainer_type), target, intent(inout) :: diag_axis(:)    !< Array of diag_axis objects
+    integer,                                  intent(inout) :: naxis           !< Number of axis that have been
+                                                                               !! registered
+    integer,                                  intent(inout) :: file_axis_id(:) !< The file's axis_ids
+    integer,                                  intent(inout) :: nfile_axis      !< Number of axis that have been
+                                                                               !! defined in file
+
+    class(*), pointer :: zaxis_data(:)      !< The data of the full zaxis
+    integer           :: subaxis_indices(2) !< The starting and ending indices of the subaxis relative to the full
+                                            !! axis
+    integer           :: i                  !< For do loops
+    integer           :: subaxis_id         !< The id of the new z subaxis
+    logical           :: axis_found         !< Flag that indicated if the zsubaxis already exists
+
+    !< Determine if the axis was already created
+    axis_found = .false.
+    do i = 1, nfile_axis
+      select type (axis => diag_axis(file_axis_id(i))%axis)
+      type is (fmsDiagSubAxis_type)
+        if (axis%zbounds(1) .eq. zbounds(1) .and. axis%zbounds(2) .eq. zbounds(2)) then
+          axis_found = .true.
+          subaxis_id = file_axis_id(i)
+          exit
+        endif
+      end select
+    enddo
+
+    !< Determine which of the variable's axis is the zaxis!
+    do i = 1, size(var_axis_ids)
+      select type (parent_axis => diag_axis(var_axis_ids(i))%axis)
+      type is (fmsDiagFullAxis_type)
+        if (parent_axis%cart_name .eq. "Z") then
+          !< If the axis was previously defined set the var_axis_ids and leave
+          if (axis_found) then
+            var_axis_ids(i) = subaxis_id
+            return
+          endif
+          zaxis_data => parent_axis%axis_data
+
+          select type(zaxis_data)
+          type is (real(kind=r4_kind))
+            !TODO need to include the conversion to "real" because nearest_index doesn't take r4s and r8s
+            subaxis_indices(1) = nearest_index(real(zbounds(1)), real(zaxis_data))
+            subaxis_indices(2) = nearest_index(real(zbounds(2)), real(zaxis_data))
+          type is (real(kind=r8_kind))
+            subaxis_indices(1) = nearest_index(real(zbounds(1)), real(zaxis_data))
+            subaxis_indices(2) = nearest_index(real(zbounds(2)), real(zaxis_data))
```

Comment:
Once axis_utils2 is updated for mixed mode, the conversion to real can be done correctly:
```F90
          type is (real(kind=r4_kind))
            !TODO need to include the conversion to "real" because nearest_index doesn't take r4s and r8s
            subaxis_indices(1) = nearest_index(zbounds(1), zaxis_data)
            subaxis_indices(2) = nearest_index(zbounds(2), zaxis_data)
          type is (real(kind=r8_kind))
            subaxis_indices(1) = nearest_index(real(zbounds(1), kind=r8_kind), zaxis_data)
            subaxis_indices(2) = nearest_index(real(zbounds(2), kind=r8_kind), zaxis_data)

```
Right now `nearest_index` only takes in "reals", so this won't compile ...